File: nanopay/src/net/nanopay/security/receipt/ReceiptGeneratingDAO.js
/**
 * NANOPAY CONFIDENTIAL
 *
 * [2020] nanopay Corporation
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of nanopay Corporation.
 * The intellectual and technical concepts contained
 * herein are proprietary to nanopay Corporation
 * and may be covered by Canadian and Foreign Patents, patents
 * in process, and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden unless prior written permission is obtained
 * from nanopay Corporation.
 */

foam.CLASS({
  package: 'net.nanopay.security.receipt',
  name: 'ReceiptGeneratingDAO',
  extends: 'foam.dao.ProxyDAO',

  documentation: `
    Performs receipt generation on all objects that pass through this DAO.
  `,

  imports: [
    'DAO receiptDAO'
  ],

  javaImports: [
    'foam.dao.DAO'
  ],

  properties: [
    {
      class: 'FObjectProperty',
      of: 'net.nanopay.security.receipt.ReceiptGenerator',
      name: 'generator'
    }
  ],

  methods: [
    {
      name: 'put_',
      javaCode: `
        try {
          ReceiptGenerator generator = getGenerator();
          generator.add(obj);

          // store receipt in receipt DAO
          foam.dao.DAO receiptDAO = (foam.dao.DAO) getReceiptDAO();
          receiptDAO.inX(x).put(generator.generate(obj));

          return super.put_(x, obj);
        } catch ( Throwable t ) {
          throw new RuntimeException(t);
        }
      `
    }
  ]
});

File: nanopay/src/net/nanopay/security/receipt/ReceiptGeneratingDAO.java
package net.nanopay.security.receipt;

import foam.core.FObject;
import foam.core.X;
import foam.dao.DAO;
import foam.dao.ProxyDAO;
import foam.lib.json.Outputter;
import foam.lib.json.OutputterMode;

public class ReceiptGeneratingDAO
  extends ProxyDAO
{
  protected final ReceiptGenerator generator_;

  public ReceiptGeneratingDAO(X x, ReceiptGenerator generator, DAO delegate) {
    setX(x);
    setDelegate(delegate);
    generator_ = generator;
  }

  @Override
  public FObject put_(X x, FObject obj) {
    generator_.add(obj);
    Receipt receipt = generator_.generate(obj);
    System.out.println(new Outputter(OutputterMode.STORAGE).stringify(receipt));
    return super.put_(x, obj);
  }
}
